src/datachain/data_storage/warehouse.py

```diff
@@ -76,6 +77,52 @@ def __exit__(self, exc_type, exc_value, traceback) -> None:
     def cleanup_for_tests(self):
         """Cleanup for tests."""
 
+    def _to_jsonable(self, obj: Any) -> Any:
+        """Recursively convert Python/Pydantic structures into JSON-serializable
+        objects.
+
+        Rules:
+        - Pydantic models -> model_dump() dict
+        - dict -> keys coerced to strings, values converted recursively
+          - Pydantic or complex keys are JSON-encoded to a stable string
+          - Primitive keys (str/int/float/bool/None) are stringified when needed
+        - list/tuple/set -> list with items converted recursively
+        - numpy arrays are expected to be converted to list earlier in convert_type
+        - everything else is returned as-is
+        """
+
+        # Pydantic models
+        if ModelStore.is_pydantic(type(obj)):
+            return obj.model_dump()
+
+        # Collections
+        if isinstance(obj, dict):
+            out: dict[str, Any] = {}
+            for k, v in obj.items():
+                # Convert key to a JSON-friendly string
+                if isinstance(k, str):
+                    key_str = k
+                elif ModelStore.is_pydantic(type(k)) or isinstance(
+                    k, (dict, list, tuple, set)
+                ):
+                    try:
+                        key_str = json.dumps(self._to_jsonable(k), ensure_ascii=False)
+                    except Exception:  # noqa: BLE001
+                        key_str = str(k)
+                elif isinstance(k, (int, float, bool)) or k is None:
+                    key_str = str(k)
+                else:
+                    key_str = str(k)
```

Comment:
This can be simplified IMO:
```suggestion
                else:
                    key_str = str(k)
```